### PR TITLE
Feature/field restrictions

### PR DESCRIPTION
**Pull request by furkansenharputlu**
Fixes https://github.com/TykTechnologies/tyk/issues/3017

### CODE REVIEW

File: gateway/api_definition.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ import (
 	"text/template"
 	"time"
 
+	"github.com/jensneuse/graphql-go-tools/pkg/graphql"
```
Comment (spetrunin) [inline gateway/api_definition.go:20]: could we group imports?
like
```
import (
  std_imports
 
  third party

  tyk
)
```

File: gateway/middleware.go [hunk 2]
```diff
@@ -403,6 +403,20 @@ func (t BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
 							}
 						}
 
+						for _, t := range v.RestrictedTypes {
```
Comment (spetrunin) [inline gateway/middleware.go:406]: do we support partitioning of restricted types?

Comment (spetrunin) [reply gateway/middleware.go:406]: I know it is unrelated to this PR but could we make iteration variables names more meaningful?

like `v -> accessRight` and `k -> apiID`

File: gateway/mw_granular_access.go [hunk 3]
```diff
@@ -30,34 +32,56 @@ func (m *GranularAccessMiddleware) ProcessRequest(w http.ResponseWriter, r *http
...
+			w.Header().Set(headers.ContentType, headers.ApplicationJSON)
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = result.Errors.WriteResponse(w)
+			m.Logger().Errorf("Error during GraphQL request restricted fields validation: '%s'", result.Errors)
+			return errCustomBodyResponse, http.StatusBadRequest
+		}
+
+		return nil, http.StatusOK
+
+	} else {
```
Comment (spetrunin) [inline gateway/mw_granular_access.go:58]: `else` branch is not required here as you already have returns in `if`

File: gateway/middleware.go (hunk 2 above)
Comment (furkansenharputlu) [reply gateway/middleware.go:406]: You are right, it is very unclear but better to make it in another PR.

**Pull request merged**